conda_libmamba_solver/solver.py

```diff
@@ -256,18 +257,43 @@ def _spinner_msg_solving(self):
                 return "Getting pinned dependencies"
         return "Solving environment"
 
+    def _max_attempts(self, in_state: SolverInputState):
+        from_env_var = os.environ.get("CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS")
+        n_installed = len(in_state.installed)
+        if from_env_var:
+            try:
+                max_attempts_from_env = int(from_env_var)
+            except ValueError:
+                raise CondaValueError(
+                    f"CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='{from_env_var}'. Must be int."
+                )
+            if max_attempts_from_env < 1:
+                raise CondaValueError(
+                    f"CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='{max_attempts_from_env}'. Must be >=1."
+                )
+            if max_attempts_from_env > n_installed:
+                log.warning(
+                    "CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='%s' is higher than the number of "
+                    "installed packages (%s) and will be ignored.",
+                    max_attempts_from_env,
+                    n_installed,
+                )
+                return n_installed
+            return max_attempts_from_env
+        if in_state.update_modifier.FREEZE_INSTALLED:
+            # this the default, but can be overriden with --update-specs
+            # TODO: should we cap this at a reasonable number? some base envs have 100s of pkgs
```

Comment:
This is the main point of this PR :) Thoughts?

Reply:
I haven't messed with `libmamba` solver at all, so absolutely no idea/feeling about what ranges would be sensible.

If you don't want a hard limit, you could use some function that just increases slower for higher inputs.
If you want a hard limit you could either do `min(n_installed, limit)` or something that only approaches the limit gradually like `math.ceil(n_installed / (1 + (n_installed / limit)**2)**.5)` or something more/less sophisticated.

Reply:
Each solve is usually sub-second, but some complex one that need some extra backtracking (like the one reported in your issue) take a 2-4s. If you end up with 100 installed packages then you need to wait a few minutes for the solver to give up.

By "giving up" I mean that we stop trying to unlock installed packages and we let the solver modify them as needed, just as mamba does. So maybe we can get by with 10 attempts. In other words, let the solver unfreeze up to 5-10 conflicting installed packages one by one until we just let everything float.

Alternatively, we could make the retry loop be based on time spent and not iterations, but that might be more complex than necessary.

Reply:
I'm going to try with 10 attempts and see if that is enough to pass all tests. Then we adjust as necessary.

Reply:
Makes sense.
Could you add that as a constant variable (`DEFAULT_LIBMAMBA_SOLVER_MAX_ATTEMPTS` or something ) to better track that?

----
> Alternatively, we could make the retry loop be based on time spent and not iterations, but that might be more complex than necessary.

Yes, complexity, but also determinism ;).